### PR TITLE
Minimalistic support for group filtering in oidc connector

### PR DESCRIPTION
**Pull request by mdpradeep**
#### Overview

Filter users based on groups in OIDC connector.

#### What this PR does / why we need it
Currently it is not possible to filter users based on groups like it is done with the `google` connector. This PR is an attempt to support group filtering in the `oidc` connector. The connector configuration takes an additional configuration as below to specify what the allowed groups are. If the user's groups belong to at least one of the `allowedGroups`, then authentication will be successful. The other change is to indicate that the user does not belong to allowed groups if authentication fails because the user does not belong to any of the allowed groups.

```
{
            "name": "My-oidc",
            "config": {
                "issuer": "<ISSUER>",
                "clientID": "12345678",
                "redirectURI": "<Dex Callback>",
                "clientSecret": "12345678",
                "insecureSkipEmailVerified": true,
                "insecureEnableGroups": true,
                "getUserInfo": true,
                "allowedGroups": ["XYZ", "ABC"],
                "scopes": ....
            },
            "id": "myoidc",
            "type": "oidc"
        }
```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: connector/oidc/oidc.go [hunk 1]
```diff
@@ -341,6 +345,7 @@ func (c *oidcConnector) createIdentity(ctx context.Context, identity connector.I
 		return identity, errors.New("oidc: no id_token in token response")
 	}
 
+	c.logger.Debugf("Claims::: %q \n", claims)
```
Comment (sagikazarmark) [inline connector/oidc/oidc.go:348]: I assume this stayed here by accident?

File: connector/oidc/oidc.go [hunk 2]
```diff
@@ -351,6 +356,7 @@ func (c *oidcConnector) createIdentity(ctx context.Context, identity connector.I
 		if err := userInfo.Claims(&claims); err != nil {
 			return identity, fmt.Errorf("oidc: failed to decode userinfo claims: %v", err)
 		}
+		c.logger.Debugf("UserInfo::: %q \n", userInfo)
```
Comment (sagikazarmark) [inline connector/oidc/oidc.go:359]: I assume this stayed here by accident?

File: connector/oidc/oidc.go [hunk 3]
```diff
@@ -51,6 +52,7 @@ type Config struct {
 
 	// InsecureEnableGroups enables groups claims. This is disabled by default until https://github.com/dexidp/dex/issues/1065 is resolved
 	InsecureEnableGroups bool `json:"insecureEnableGroups"`
+	AllowedGroups []string `json:"allowedGroups"` // allowedGroups
```
Comment (sagikazarmark) [inline connector/oidc/oidc.go:55]:
```suggestion
	AllowedGroups []string `json:"allowedGroups"`
```

File: connector/oidc/oidc.go [hunk 4]
```diff
@@ -425,6 +431,28 @@ func (c *oidcConnector) createIdentity(ctx context.Context, identity connector.I
 				}
 			}
 		}
+		c.logger.Debugf("Groups::: %q \n", groups)
```
Comment (sagikazarmark) [inline connector/oidc/oidc.go:434]: I assume this stayed here by accident?

File: connector/oidc/oidc.go [hunk 5]
```diff
@@ -425,6 +431,28 @@ func (c *oidcConnector) createIdentity(ctx context.Context, identity connector.I
...
+
+		// Validate that the user is part of allowedGroups
+		var accessAllowed = false
+		if len(c.allowedGroups) > 0 {
+			for _,group := range c.allowedGroups {
+				groupPresent := slices.Contains(groups, group)
+				if groupPresent {
+					accessAllowed = true
+					c.logger.Debugf("Found a match - Group::: %q \n", group)
```
Comment (sagikazarmark) [inline connector/oidc/oidc.go:443]: We prefer structured logging over custom message formats

File: connector/oidc/oidc.go [hunk 6]
```diff
@@ -425,6 +431,28 @@ func (c *oidcConnector) createIdentity(ctx context.Context, identity connector.I
...
+					break
+				}
+			}
+		} else {
+			// don't check for groups if allowedGroups is not configured
+			c.logger.Debugf("Group check not performed \n")
```
Comment (sagikazarmark) [inline connector/oidc/oidc.go:449]: this looks unnecessary

File: connector/oidc/oidc.go [hunk 7]
```diff
@@ -425,6 +431,28 @@ func (c *oidcConnector) createIdentity(ctx context.Context, identity connector.I
...
 			}
 		}
+		c.logger.Debugf("Groups::: %q \n", groups)
+
+		// Validate that the user is part of allowedGroups
+		var accessAllowed = false
+		if len(c.allowedGroups) > 0 {
+			for _,group := range c.allowedGroups {
+				groupPresent := slices.Contains(groups, group)
+				if groupPresent {
```
Comment (sagikazarmark) [inline connector/oidc/oidc.go:441]:
```suggestion
				if slices.Contains(groups, group) {
```

File: server/handlers.go [hunk 8]
```diff
@@ -467,7 +467,7 @@ func (s *Server) handleConnectorCallback(w http.ResponseWriter, r *http.Request)
 
 	if err != nil {
 		s.logger.Errorf("Failed to authenticate: %v", err)
-		s.renderError(r, w, http.StatusInternalServerError, fmt.Sprintf("Failed to authenticate: %v", err))
+		s.renderError(r, w, http.StatusUnauthorized, fmt.Sprintf("Failed to authenticate: %v", err))
```
Comment (sagikazarmark) [inline server/handlers.go:470]: This is not a good idea. This is where we catch unexpected errors as well. If there is a need to return 401 based on the error, the error needs to be checked for such conditions.

**Review by sagikazarmark — CHANGES_REQUESTED** [review]

File: connector/oidc/oidc.go (hunk 1 above)
Comment (mdpradeep) [reply connector/oidc/oidc.go:348]: Actually this came as a need when we were recently debugging stuff on a production environment where we did not have access to the IdP. It became painful to prove that some of the claims that we were expecting were not being sent by the IdP. But if you think it is unnecessary I will remove it.

File: connector/oidc/oidc.go (hunk 2 above)
Comment (mdpradeep) [reply connector/oidc/oidc.go:359]: Similar to  the previous need, it was helpful to see to which groups the user belonged to. Again, I will remove it if you think it is not necessary.

File: connector/oidc/oidc.go (hunk 4 above)
Comment (mdpradeep) [reply connector/oidc/oidc.go:434]: This is redundant, will remove it.

File: connector/oidc/oidc.go (hunk 5 above)
Comment (mdpradeep) [reply connector/oidc/oidc.go:443]: Sure, could you point me to an example?

File: connector/oidc/oidc.go (hunk 6 above)
Comment (mdpradeep) [reply connector/oidc/oidc.go:449]: Will remove it.

File: server/handlers.go (hunk 8 above)
Comment (mdpradeep) [reply server/handlers.go:470]: Completely agree. I was not intending to add this to  the PR actually. So, are you suggesting we create a custom Error  type with code + error message and then check for `code` to decide whether to return `StatusInternalServerError` or `StatusUnauthorized` or are you saying that we check for the error message string and then decide whether to return `StatusInternalServerError` or `StatusUnauthorized`?

Comment (nabokihms) [reply server/handlers.go:470]: Let's just revert this change. From the PR description, I don't think this is required.

File: connector/oidc/oidc.go [hunk 9]
```diff
@@ -425,6 +431,27 @@ func (c *oidcConnector) createIdentity(ctx context.Context, identity connector.I
...
+			}
+		} else {
+			// don't check for groups if allowedGroups is not configured
+			c.logger.Debugf("Group check not performed \n")
+			accessAllowed = true
+		}
+
+		if !accessAllowed {
+			return identity, errors.New("User is not in the allowed group(s)")
+		}
```
Comment (nabokihms) [inline connector/oidc/oidc.go:454]:
```suggestion
		if len(c.allowedGroups) > 0 {
			groupMatches := groups.Filter(ident.Groups, p.allowedGroups)

			if len(groupMatches) == 0 {
				// No group membership matches found, disallowing
				return identity, fmt.Errorf("user not a member of allowed groups")
			}
			
			groups = groupMatches
		}
```

https://github.com/dexidp/dex/blob/master/pkg/groups/groups.go Should be used here to be inline with other connectors.

**Review by nabokihms — CHANGES_REQUESTED** [review]
Debug messages don't seem like a necessary addition. The better approach is, for example, to trace HTTP requests to providers and check returned bodies. Custom debug messages are blocking this PR.

File: connector/oidc/oidc.go (hunk 1 above)
Comment (mdpradeep) [reply connector/oidc/oidc.go]: Removed this in 345eba905ddb2664855e79da2acf816ad6b97eb5

File: connector/oidc/oidc.go (hunk 2 above)
Comment (mdpradeep) [reply connector/oidc/oidc.go]: Removed this in 345eba905ddb2664855e79da2acf816ad6b97eb5

File: connector/oidc/oidc.go (hunk 4 above)
Comment (mdpradeep) [reply connector/oidc/oidc.go]: Removed in 345eba905ddb2664855e79da2acf816ad6b97eb5

File: connector/oidc/oidc.go (hunk 5 above)
Comment (mdpradeep) [reply connector/oidc/oidc.go]: Handled in 14f867af2fa25cd4a440c7779431223ab55796af

File: connector/oidc/oidc.go (hunk 6 above)
Comment (mdpradeep) [reply connector/oidc/oidc.go]: Handled in 14f867af2fa25cd4a440c7779431223ab55796af

File: server/handlers.go (hunk 8 above)
Comment (mdpradeep) [reply server/handlers.go]: Done in 345eba905ddb2664855e79da2acf816ad6b97eb5

**Comment by nabokihms** [comment]
@mdpradeep, everything seems fine now according to the code and ready to be merged (when tests are successfully passed).

 I'd like to take this feature to the upcoming v2.38.0 release, so waiting 🙂

File: connector/oidc/oidc.go (hunk 9 above)
Comment (nabokihms) [reply connector/oidc/oidc.go]: waiting for this one, because it was lost on rebasing

Comment (mdpradeep) [reply connector/oidc/oidc.go]: Fixed in https://github.com/dexidp/dex/pull/3074/commits/b3e5bf544bbc192d6013f4b361c597ad32a4bd4c

**Review by nabokihms — APPROVED** [review]
Looks really nice. Thank you, @mdpradeep, for your contribution!

**Pull request merged**